joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/passes/FullNameUniquenessPass.scala

```diff
@@ -0,0 +1,134 @@
+  * duplicates, and appends unique suffixes to ensure each one can be uniquely identified. Calls to static methods
+  * affected by this are also updated to reflect the new method fullName.
+  */
+class FullNameUniquenessPass(cpg: Cpg) extends CpgPass(cpg) {
+
+  private type AffectedNodeType = Method | TypeDecl | NamespaceBlock
+  private val logger = LoggerFactory.getLogger(classOf[FullNameUniquenessPass])
+
+  override def run(dstGraph: DiffGraphBuilder): Unit = {
+    handleMethods(dstGraph)
+    handleTypeDecls(dstGraph)
+    handleNamespaceBlocks(dstGraph)
+  }
+
+  private def handleMethods(dstGraph: DiffGraphBuilder): Unit = {
+    val methodFullNameMap = generateStableFullNameMapping(cpg.method.nameNot(NamespaceTraversal.globalNamespaceName))
+    methodFullNameMap.foreach { case (fullName, methods) =>
+      logDuplicates(methods, fullName)
+      val sortedMethods = sortNodesByLocation(methods)
+      lazy val callsAffected =
+        cpg.call.methodFullNameExact(fullName).filterNot(_.file.exists(_.name == sortedMethods.head.filename)).l
+      lazy val bindingsAffected = cpg.binding.methodFullNameExact(fullName).l
+      sortedMethods.tail.zipWithIndex.foreach { case (method, index) =>
+        val signature   = method.signature
+        val isCFunction = !fullName.endsWith(s":$signature")
+        val newFullName = if (isCFunction) {
+          s"$fullName${Defines.DuplicateSuffix}$index"
+        } else {
+          val fullNameWithoutSignature = fullName.stripSuffix(s":$signature")
+          s"$fullNameWithoutSignature${Defines.DuplicateSuffix}$index:$signature"
+        }
+        dstGraph.setNodeProperty(method, Method.PropertyNames.FullName, newFullName)
+        // fixup bindings
+        bindingsAffected.filter(_.refOut.contains(method)).foreach { binding =>
+          dstGraph.setNodeProperty(binding, Binding.PropertyNames.MethodFullName, newFullName)
+        }
+        // fixup calls to static methods in the same compilation unit via the naive namespace-by-filename approach
+        if (method.isStatic.nonEmpty) {
+          val callCandidates = callsAffected.filter(_.file.exists(_.name == method.filename))
+          callCandidates.foreach { call =>
+            dstGraph.setNodeProperty(call, Call.PropertyNames.MethodFullName, newFullName)
+          }
+        }
+      }
+    }
+  }
+
+  private def generateStableFullNameMapping[T <: AffectedNodeType](nodes: Iterator[T]): Seq[(String, List[T])] = {
+    nodes
+      .groupBy(_.fullName)
+      .filter(_._2.size > 1)
+      // sort for stable output
+      .sortBy(_._1)
+  }
+
+  private def sortNodesByLocation[T <: AffectedNodeType](nodes: List[T]): List[T] = {
+    nodes.sortBy { node =>
+      val fileName = node.filename
+      val lineNr   = node.lineNumber.getOrElse(-1)
+      val columnNr = node.columnNumber.getOrElse(-1)
+      s"$fileName:$lineNr:$columnNr"
```

Comment:
I think sorting by a tuple should also work, and it avoids problems like `10 < 2` in lexicographical sorting
```suggestion
      (fileName, lineNr, columnNr)
```